File: compiler/bench/heapsort.fs
\ expected: 49995000
\ Heapsort benchmark - sort 10000 elements, sum result

10000 constant SIZE
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: swap-arr ( i j -- )
  over arr@ over arr@
  rot arr! swap arr! ;

: sift-down ( start end -- )
  begin
    dup 2* 1+           \ start end child
    2dup >= if drop 2drop exit then
    dup 1+ 2 pick <= if
      dup arr@ over 1+ arr@ < if 1+ then
    then
    dup arr@ 2 pick arr@ >= if drop 2drop exit then
    2 pick over swap-arr
    rot drop swap
  again ;

: heapify ( -- )
  SIZE 2 / 1-
  begin dup 0>= while
    dup SIZE 1- sift-down
    1-
  repeat drop ;

: hsort ( -- )
  heapify
  SIZE 1-
  begin dup 0> while
    0 over swap-arr
    0 over 1- sift-down
    1-
  repeat drop ;

: init-arr ( -- )
  SIZE 0 do
    SIZE i - 1- i arr!
  loop ;

: sum-arr ( -- n )
  0 SIZE 0 do i arr@ + loop ;

: main
  init-arr
  hsort
  sum-arr . cr ;

File: compiler/bench/heapsort.c
// expected: 49995000
// Heapsort benchmark - sort 10000 elements, sum result

#include <stdio.h>

#define SIZE 10000

int arr[SIZE];

void swap(int i, int j) {
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

void sift_down(int start, int end) {
    while (1) {
        int child = 2 * start + 1;
        if (child > end) return;
        if (child + 1 <= end && arr[child] < arr[child + 1]) {
            child++;
        }
        if (arr[start] >= arr[child]) return;
        swap(start, child);
        start = child;
    }
}

void heapify(void) {
    for (int start = SIZE / 2 - 1; start >= 0; start--) {
        sift_down(start, SIZE - 1);
    }
}

void hsort(void) {
    heapify();
    for (int end = SIZE - 1; end > 0; end--) {
        swap(0, end);
        sift_down(0, end - 1);
    }
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = SIZE - i - 1;
    }
}

long sum_arr(void) {
    long sum = 0;
    for (int i = 0; i < SIZE; i++) {
        sum += arr[i];
    }
    return sum;
}

int main(void) {
    init_arr();
    hsort();
    printf("%ld\n", sum_arr());
    return 0;
}
